vane-trifles/src/main/java/org/oddlama/vane/trifles/StorageGroup.java

```diff
@@ -138,11 +149,56 @@ public void on_inventory_click(InventoryClickEvent event) {
                 break;
         }
 
+        switch (event.getClick()) {
+            case NUMBER_KEY:
+                // Deny swapping storage items with number keys into storage inventory, but allow in player inventory
+                var swapped_item_is_storage = is_storage_item(player.getInventory().getItem(event.getHotbarButton()));
+                cancel = (swapped_item_is_storage || clicked_item_is_storage) && !clicked_inventory_is_player_inventory;
+                break;
+        }
+
         if (cancel) {
             event.setCancelled(true);
         }
     }
 
+    @EventHandler(priority = EventPriority.HIGH, ignoreCancelled = true)
+    public void on_drop_item(PlayerDropItemEvent event) {
+        if (!is_storage_item(event.getItemDrop().getItemStack())) {
+            return;
+        }
+
+        // Close the inventory if the player drops the currently open storage item
+        var storage_item_is_open_state = is_currently_open(event.getItemDrop().getItemStack());
+        if (storage_item_is_open_state) {
+            var is_known_custom_inventory = open_block_state_inventories.containsKey(event.getPlayer().getOpenInventory().getTopInventory());
+            if (is_known_custom_inventory) {
+                event.getPlayer().closeInventory(InventoryCloseEvent.Reason.CANT_USE);
+            } else {
+                // Item shouldn't be tagged as open if a custom inventory is not open, fix open tag
+                event.getItemDrop().getItemStack().editMeta(meta -> {
+                    meta.getPersistentDataContainer().set(STORAGE_IS_OPEN, PersistentDataType.BOOLEAN, false);
+                });
+            }
+        }
+    }
+
+    @EventHandler(priority = EventPriority.HIGH, ignoreCancelled = true)
+    public void on_pickup_item(EntityPickupItemEvent event) {
+        if (!(event.getEntity() instanceof Player player)) {
+            return;
+        }
+
+        if (!is_storage_item(event.getItem().getItemStack())) {
+            return;
+        }
+
+        // Ensure bugged/old storage items are set to closed when picked up just in case
+        event.getItem().getItemStack().editMeta(meta -> {
+            meta.getPersistentDataContainer().set(STORAGE_IS_OPEN, PersistentDataType.BOOLEAN, false);
+        });
```

Comment:
Very nice, I was just about to ask what happens when the client disconnects while having a storage item opened, or when a server crashes occurs.

Altough I'm not entirely sure whether it will be possible to move or drop bugged items at all times.